modules/csg/csg.h

```diff
@@ -138,99 +140,59 @@ struct CSGBrushOperation {
-			int points[2];
-			Edge() {
-				outer = false;
-			}
+		struct Face2D {
+			int vertex_idx[3];
 		};
 
-		Vector<Edge> edges;
-		Ref<Material> material;
-		bool smooth;
-		bool invert;
-
-		int base_edges; //edges from original triangle, even if split
-
-		void _clip_segment(const CSGBrush *p_brush, int p_face, const Vector2 *segment, MeshMerge &mesh_merge, bool p_for_B);
-
-		void create(const CSGBrush *p_brush, int p_face, MeshMerge &mesh_merge, bool p_for_B);
-		void clip(const CSGBrush *p_brush, int p_face, MeshMerge &mesh_merge, bool p_for_B);
-	};
-
-	struct PolyPoints {
-
-		Vector<int> points;
-
-		Vector<Vector<int> > holes;
-	};
-
-	struct EdgeSort {
-		int edge;
-		int prev_point;
-		int edge_point;
-		float angle;
-		bool operator<(const EdgeSort &p_edge) const { return angle < p_edge.angle; }
+		Vector<Vertex2D> vertices;
+		Vector<Face2D> faces;
+		Plane plane;
+		Transform to_2D;
+		Transform to_3D;
+		float vertex_snap2;
+
+		inline int _get_point_idx(const Vector2 &p_point);
+		inline int _add_vertex(const Vertex2D &p_vertex);
+		inline void _add_vertex_idx_sorted(Vector<int> &r_vertex_indices, int p_new_vertex_index);
+		inline void _merge_faces(const Vector<int> &p_segment_indices);
+		inline void _find_edge_intersections(const Vector2 p_segment_points[2], Vector<int> &r_segment_indices);
+		inline int _insert_point(const Vector2 &p_point);
+
+		void insert(const CSGBrush &p_brush, int p_brush_face);
+		void addFacesToMesh(MeshMerge &r_mesh_merge, bool p_smooth, bool p_invert, const Ref<Material> &p_material, bool p_from_b);
+
+		Build2DFaces() {}
+		Build2DFaces(const CSGBrush &p_brush, int p_brush_face, float p_vertex_snap2);
 	};
 
-	struct CallbackData {
-		const CSGBrush *A;
-		const CSGBrush *B;
-		int face_a;
-		CSGBrushOperation *self;
-		Map<int, BuildPoly> build_polys_A;
-		Map<int, BuildPoly> build_polys_B;
+	struct Build2DFaceCollection {
+		Map<int, Build2DFaces> build2DFacesA;
+		Map<int, Build2DFaces> build2DFacesB;
 	};
 
-	void _add_poly_points(const BuildPoly &p_poly, int p_edge, int p_from_point, int p_to_point, const Vector<Vector<int> > &vertex_process, Vector<bool> &edge_process, Vector<PolyPoints> &r_poly);
-	void _add_poly_outline(const BuildPoly &p_poly, int p_from_point, int p_to_point, const Vector<Vector<int> > &vertex_process, Vector<int> &r_outline);
-	void _merge_poly(MeshMerge &mesh, int p_face_idx, const BuildPoly &p_poly, bool p_from_b);
-
-	void _collision_callback(const CSGBrush *A, int p_face_a, Map<int, BuildPoly> &build_polys_a, const CSGBrush *B, int p_face_b, Map<int, BuildPoly> &build_polys_b, MeshMerge &mesh_merge);
-
-	static void _collision_callbacks(void *ud, int p_face_b);
-	void merge_brushes(Operation p_operation, const CSGBrush &p_A, const CSGBrush &p_B, CSGBrush &result, float p_snap = 0.001);
+	void update2DFaces(const CSGBrush &p_brush_a, const int p_face_idx_a, const CSGBrush &p_brush_b, const int p_face_idx_b, Build2DFaceCollection &p_collection, float p_vertex_snap);
```

Comment:
Shouldn't this be `update_2d_faces` to match our naming convention for methods?

Reply:
Old habits die hard. :-)
Looking at it, I think a better name is update_faces.